codec/src/main/java/io/netty/handler/codec/compression/CompressionHandler.java

```diff
@@ -0,0 +1,146 @@
+                              boolean discardBytesAfterFinished) {
+        this.compressorSupplier = requireNonNull(compressorSupplier, "compressorSupplier");
+        this.closeWriteTimeout = checkPositive(closeWriteTimeout, "closeWriteTimeout");
+        this.closeWriteTimeoutUnit = requireNonNull(closeWriteTimeoutUnit, "closeWriteTimeoutUnit");
+        this.discardBytesAfterFinished = discardBytesAfterFinished;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        compressor = compressorSupplier.get();
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        if (compressor != null) {
+            try {
+                finish(ctx, false);
+            } finally {
+                closeCompressor();
+            }
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        if (compressor != null) {
+            closeCompressor();
+        }
+        ctx.fireChannelInactive();
+    }
+
+    @Override
+    public Future<Void> write(ChannelHandlerContext ctx, Object msg) {
+        if (compressor == null || !(msg instanceof ByteBuf)) {
+            return ctx.write(msg);
+        }
+        ByteBuf input = (ByteBuf) msg;
+        if (compressor.isFinished()) {
+            if (discardBytesAfterFinished) {
+                input.release();
+                return ctx.newSucceededFuture();
+            }
+            return ctx.write(msg);
+        }
+        try {
+            ByteBuf buffer = compressor.compress(input, ctx.alloc());
+            return ctx.write(buffer);
+        } finally {
+            input.release();
+        }
+    }
+
+    @Override
+    public Future<Void> close(ChannelHandlerContext ctx) {
+        return finish(ctx, true);
+    }
+
+    private Future<Void> finish(ChannelHandlerContext ctx, boolean closeCtx) {
+        if (compressor == null || compressor.isFinished()) {
+            return closeCtx ? ctx.close() : ctx.newSucceededFuture();
+        }
+        ByteBuf buffer = compressor.finish(ctx.alloc());
+        if (!buffer.isReadable()) {
+            buffer.release();
+            return closeCtx ? ctx.close() : ctx.newSucceededFuture();
+        }
+        if (closeCtx) {
+            Promise<Void> promise = ctx.newPromise();
+            Future<Void> f = ctx.writeAndFlush(buffer).addListener(ctx, (c, ignore) -> c.close().cascadeTo(promise));
+            if (!f.isDone()) {
+                // Ensure the channel is closed even if the write operation completes in time.
+                Future<?> sF =  ctx.executor().schedule(() -> ctx.close().cascadeTo(promise),
+                        closeWriteTimeout, closeWriteTimeoutUnit);
+                f.addListener(sF, (scheduledFuture, ignore) -> scheduledFuture.cancel());
+            }
+            return promise.asFuture();
```

Comment:
This is what I meant:

```suggestion
            Future<Void> f = ctx.writeAndFlush(buffer).flatMap(ignored -> ctx.close());
            if (!f.isDone()) {
                // Ensure the channel is closed even if the write operation completes in time.
                Future<?> sF =  ctx.executor().schedule(() -> {}, closeWriteTimeout, closeWriteTimeoutUnit)
                        .flatMap(ignored -> ctx.close());
                f.addListener(sF, (scheduledFuture, ignore) -> scheduledFuture.cancel());
            }
            return f;
```

Reply:
Let me pull it in as it is and then thing again about if the flap map is correct :) Just so I can make some progress